### ISSUE TITLE
Publishing Entities causes  serious performance issues

### ISSUE THREAD

**Issue by laceysanderson** [tripal/tripal#530]
### System information

* Tripal Version: 7.x-3.x-dev
* Drupal Version: 7.x
* PostgreSQL Version: n/a
* PHP Version: n/a

### Issue description

Currently publishing entities can cause a site to become unusable due to excessive slowness -it can even cause timeouts. Furthermore, when publishing a large number of entities it can cause continuous timeouts making the site completely unavailable. For example, I recently tried to publish my SNP pages in KnowPulse (~426K) and after running for 15hrs the site become completely unresponsive for >2hrs requiring me to kill the job rolling back all progress.

#### Steps to reproduce

1. Go to Admin > Structure > Tripal Content Types and click "Publish Tripal Content"
2. Fill out the drop-down to select a content type for which you have a large amount of data and submit the job.

#### Error messages and screenshots

There are no error messages or screenshots for this issue since it's performance rather then errors that are the problem.

**Comment by laceysanderson** [comment]
My first step on this is going to be to allow the admin to specify in the publish form to only publish a given number of entities. We had this feature in Tripal2 and it can be used to effectively chunk the publishing process into more reasonable length transactions.

Other **potential optimizations** based on looking at the code:
- Optimize, estimate or remove the count query used to report progress. We can't show the progress through the UI anyway due to the transaction so how useful is this?
- Make a way to create multiple entities at once to take advantage of bulk inserts rather then one at a time.
- Attempt to optimize query built to select chado records to publish

I also need to evaluate the transaction more. Long running transactions can have serious performance issues in PostgreSQL due to the ACID-compliance keeping those records cordoned off until the transaction is committed --you should have seen the christmas lights in htop when I had to cancel my publish job ;-p). One option is to move the transaction within the loop so that each one is not very long but then we end up with more overhead creating transactions. A better albeit more complicated approach might be to break the job into chunks where each one is a single transaction. I'll do some timing on the various options as far as this goes.

**Comment by laceysanderson** [comment]
So interestingly, I didn't actually do any of the potential optimizations above. Instead, once I started profiling where the time was spent, I found a lot of re-loading of the same objects -removing this sped things up drastically. Furthermore, there was a  good speed increase once I broke the job into chunks where each one is a single transaction. Combined these two optimizations were enough for my purposes and thus are submitted in the above PR.
